archive/compression/compression_test.go

```diff
@@ -72,12 +76,113 @@ func testCompressDecompress(t *testing.T, size int, compression Compression) {
 	if !bytes.Equal(orig, decompressed) {
 		t.Fatal("strange decompressed data")
 	}
+
+	return decompressor
 }
 
 func TestCompressDecompressGzip(t *testing.T) {
-	testCompressDecompress(t, 1024*1024, Gzip)
+	os.Setenv(disablePgzipEnv, "1")
+	defer os.Unsetenv(disablePgzipEnv)
+
+	decompressor := testCompressDecompress(t, 1024*1024, Gzip)
+
+	wrapper := decompressor.(*readCloserWrapper)
+	_, ok := wrapper.Reader.(*gzip.Reader)
+	if !ok {
+		t.Fatalf("unexpected compressor type: %T", wrapper.Reader)
+	}
+}
+
+func TestCompressDecompressPgzip(t *testing.T) {
+	decompressor := testCompressDecompress(t, 1024*1024, Gzip)
+
+	wrapper := decompressor.(*readCloserWrapper)
+	_, ok := wrapper.Reader.(*pgzip.Reader)
+	if !ok {
+		t.Fatalf("unexpected compressor type: %T", wrapper.Reader)
+	}
 }
 
 func TestCompressDecompressUncompressed(t *testing.T) {
 	testCompressDecompress(t, 1024*1024, Uncompressed)
 }
+
+func TestDisablePgzip(t *testing.T) {
+	os.Setenv(disablePgzipEnv, "1")
+	if !disablePgzip() {
+		t.Errorf("disable via %s doesn't work", disablePgzipEnv)
+	}
+
+	os.Unsetenv(disablePgzipEnv)
+	if disablePgzip() {
+		t.Error("pzsip should not be disabled by default")
+	}
+}
+
+func generateCompressedData(b *testing.B, sizeInMb int) []byte {
+	size := sizeInMb * 1024 * 1024
+	b.SetBytes(int64(size))
+
+	// Make results predictable
+	rand.Seed(1234)
+
+	data := make([]byte, size)
+	if _, err := rand.Read(data); err != nil {
+		b.Fatal(err)
+	}
+
+	var buf bytes.Buffer
+	compressor, err := CompressStream(&buf, Gzip)
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	if n, err := compressor.Write(data); err != nil || n != size {
+		b.Fatal(err)
+	}
+
+	compressor.Close()
+	return buf.Bytes()
+}
+
+func benchmarkDecompression(b *testing.B, sizeInMb int) {
+	compressed := generateCompressedData(b, sizeInMb)
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		decompressor, err := DecompressStream(bytes.NewReader(compressed))
+		if err != nil {
+			b.Fatal(err)
+		}
+
+		_, err = ioutil.ReadAll(decompressor)
+		if err != nil {
+			b.Fatal(err)
+		}
+
+		decompressor.Close()
+	}
+}
+
+func benchmarkGzipDecompression(b *testing.B, sizeInMb int) {
+	os.Setenv(disablePgzipEnv, "1")
+	defer os.Unsetenv(disablePgzipEnv)
+
+	benchmarkDecompression(b, sizeInMb)
+}
+
+func benchmarkPgzipDecompression(b *testing.B, sizeInMb int) {
+	benchmarkDecompression(b, sizeInMb)
+}
+
+func BenchmarkGzipDecompression64(b *testing.B)  { benchmarkGzipDecompression(b, 64) }
```

Comment:
The benchmarks can just be `BenchmarkDecompression*` and run with 2 different environments. This also makes it easier to use with `benchcmp`